tarpc/src/client.rs

```diff
@@ -41,13 +41,26 @@ pub struct Config {
     /// `pending_requests_buffer` controls the size of the channel clients use
     /// to communicate with the request dispatch task.
     pub pending_request_buffer: usize,
+
+    /// An implementation of RequestSequencer, to provide a unique series of request ids.
+    /// The default implementation generates 0,1,2,3,4,5,..., but this option can be leveraged
+    /// to generate less predictable results, using a block cipher for example.
+    pub request_sequencer: Arc<dyn RequestSequencer>
```

Comment:
I think the problem here is that you have set up multiple tarpc clients, but on the server side you have them all multiplexed on a single server channel. tarpc server channels only support a single client. The right way to do this would be to have multiple server channels all using the same underlying mqtt transport. That way, request IDs can overlap on the transport, but they are unique when they reach the server channel.

I think this is a fundamental difference between MQTT—which natively supports many-to-one relationships between client and server—and tarpc. I think we need to bridge the gap here, but I *don't* think we should do that by changing tarpc to support many-to-one server channels.

Does the MQTT protocol even have a way for the server to know if a message came from a different client? It seems like the entire notion of a client is kind of made up in this case. In MQTT, it's just seeing a client publishing a message with a different response topic, right? But for tarpc-over-MQTT, we need a way to uniquely identify clients. Can we use the correlation data for this purpose? I feel like something needs to map correlation data to unique clients, and to me, the transport feels most appropriate here.

Reply:
`Does the MQTT protocol even have a way for the server to know if a message came from a different client? It seems like the entire notion of a client is kind of made up in this case. In MQTT, it's just seeing a client publishing a message with a different response topic, right? But for tarpc-over-MQTT, we need a way to uniquely identify clients. Can we use the correlation data for this purpose? I feel like something needs to map correlation data to unique clients, and to me, the transport feels most appropriate here.`

No it doesn't. MQTT doesn't really have the notion of server and client either. Well MQTTv5 specs out request-response, but it's not really aimed at server and client, they are more or less equal peers. 
What we CAN do is to define on the tarpc level that one needs to publish requests to /request/<clientid>. That way the server can use the topic to distinguish between clients. But I think we can get away with this on the transport level by simply modifying the request ids by some cryptography / hashing to map the (ClientId, RequestId) into a globally unique RequestId. It is not a perfect solution, but it should be unique enough in practice. I don't see an easy way to define a separate channel for each client.

About correlation data: again, it can be meaningful for the requestor side only, to pair up a response with the original request. The server needs to be able to respond to any correlation data without inspecting it, including the case where all clients always generate 0x0000000000 for the CD.

I think we discussed the matter enough for me to be able to prepare another version of this PR, give me a week to get to it. We're mid-renovations, so it's a bit hard to get some free time.

Reply:
> one needs to publish requests to /request/

Would this be a topic prefix?

> I don't see an easy way to define a separate channel for each client.

I was thinking the transport could be responsible for establishing new connections based on a /client/<uid>/ prefix to the topic. I threw together a prototype [here](https://github.com/tikue/tarpc-mqtt-transport/blob/master/src/server_transport/dispatch.rs).

> About correlation data: again, it can be meaningful for the requestor side only

Right, of course!

> We're mid-renovations, so it's a bit hard to get some free time.

No worries at all, there's absolutely no time pressure on any of this :)

Reply:
Hmm, I effed something up with the rebases and merges, so I'll need to clean this up, but I have a next preliminary version that adds the request context to the Request and Response structs.

I couldn't get it working by saving the extensions into `Context::current()`. Seems like that is always cloned (the span::Current already is), and if I mutate it, I'm actually mutating only a clone of it, so it gets lost.
If there were a way to do that (just thought about internal mutability perhaps?) we could get rid of the extensions from the Request and Response structs and rely on the span context instead.